rzup/src/settings.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2025 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use crate::env::Environment;
+use crate::error::Result;
+use crate::events::RzupEvent;
+use crate::RzupError;
+
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs;
+
+#[derive(Debug, Serialize, Deserialize, Default, Clone)]
+pub struct Settings {
+    #[serde(default)]
+    active_versions: HashMap<String, String>,
+}
+
+impl Settings {
+    pub(crate) fn load(env: &Environment) -> Result<Self> {
+        let settings_path = env.settings_path();
+        if !settings_path.exists() {
+            env.emit(RzupEvent::Debug {
+                message: format!(
+                    "No settings file found at {}, using defaults",
+                    settings_path.display()
+                ),
+            });
+            return Ok(Self::default());
+        }
+
+        env.emit(RzupEvent::Debug {
+            message: format!("Loading settings from {}", settings_path.display()),
+        });
+        let contents = fs::read_to_string(settings_path)
+            .map_err(|e| RzupError::Environment(format!("Failed to read settings file: {}", e)))?;
+
+        toml::from_str(&contents)
+            .map_err(|e| RzupError::Environment(format!("Failed to parse settings file: {}", e)))
+    }
+
+    pub(crate) fn save(&self, env: &Environment) -> Result<()> {
+        if let Some(parent) = env.settings_path().parent() {
+            fs::create_dir_all(parent).map_err(|e| {
+                RzupError::Environment(format!("Failed to create settings directory: {}", e))
+            })?;
+        }
+
+        env.emit(RzupEvent::Debug {
+            message: format!("Saving settings to {}", env.settings_path().display()),
+        });
+
+        let contents = toml::to_string_pretty(self)
+            .map_err(|e| RzupError::Environment(format!("Failed to serialize settings: {}", e)))?;
```

Comment:
I believe that the only way this can fail is due to a programming mistake and asserts seem appropriate to me in that situation.

Reply:
Good catch - there are a ton of infallible operations that can be refactored.